src/improc.c

```diff
@@ -2512,3 +2512,175 @@ DoubleImage ifft2DDouble(ComplexImage image) {
+  return quack->buffer[quack->start];
+}
+
+static int quackPopFront(Quack *quack) {
+  int value = quackPeekFront(quack);
+  quack->start = (quack->start + 1) % quack->capacity;
+  quack->size--;
+  return value;
+}
+
+#if 0
+// this function is unused, but i don't know if we want to keep it for completion sake
+static void quackPushFront(Quack *quack, int value) {
+  if (quack->size >= quack->capacity) {
+    fatalError("attempted to insert into full quack. capactiy: %d", quack->capacity);
+  }
+  quack->start = (quack->start + quack->capacity - 1) % quack->capacity;
+  quack->buffer[quack->start] = value;
+  quack->size++;
+}
+#endif
+
+static Quack createNewQuackWithMemory(int capacity, int *memory) {
+  Quack quack;
+  quack.buffer = memory;
+  quack.start = 0;
+  quack.end = 0;
+  quack.size = 0;
+  quack.capacity = capacity;
+
+  return quack;
+}
+
+/**
+* Compute a sliding window (of size `w`) minimum/maximum depending on the `ord` flag of the 1D array (size = n).
+* The `offset` and `start` parameters are used to change the indexing in the array so that element i is accessed at
+* location i * offset + start. A `workspace` array (of size w elements) is given (does not need to be initialised) to
+* use as memory for the quack implementation.
+*
+* @param img The 1D int array that the sliding window operation will be computed on
+* @param out The 1D int array that the result of the operation will be stored in
+* @param n The size of both of the arrays (in nr of elements)
+* @param w The size of the window
+* @param ord If `ord` == 1, the sliding minimum is computed. Otherwise a sliding maximum is computed
+* @param offset Determines the gap in position in the input/output arrays between two successive elements
+* @param start Determines the position of the first element in the input/output arrays
+* @param workspace An array of size w (in nr of elements) to use as backing storage for our internal datastructure.
+*/
+static void slidingWindowOrd(int *img, int *out, int n, int w, int ord, int offset, int start, int *workspace) {
+  // our quack datastructure stores indices of numers in our original array
+  Quack quack = createNewQuackWithMemory(w, workspace);
+
+  for (int i = 0; i < n; i++) {
+    // remove any elements from the front (oldest side) of the quack that are now out of the "reach" of our window
+    while (!quackIsEmpty(&quack) && quackPeekFront(&quack) <= i - w) {
+      quackPopFront(&quack);
+    }
+
+    // we are now observing the current value, img[i]. we pop values from the back (most recent side) of the quack
+    // until we encounter a value that is larger (and makes the current value insignificant). our loop invariant is
+    // that we keep the largest value currently in our window at the front of the quack.
+    // NOTE: the comparison to ord is a complicated way of switching between < and >
+    while (!quackIsEmpty(&quack) && (img[quackPeekBack(&quack) * offset + start] <= img[i * offset + start]) == ord) {
```

Comment:
This line is pretty unreadable, but it has to be written like this because of the direct indexing. If we choose to transpose the image twice we can clean this up.

Reply:
Additionally, note that if ord == 1, this line says `img[quackvalue] <= img[currentvalue]` but if ord == 0 it actually says `img[quackvalue] > img[currentvalue]`. This asymmetry is not important for correctness, but we can work around it by adding an additional `|| lhs == rhs` comparison.

Reply:
Might indeed be a bit difficult to read. We can play around with transposing at some point, but for now I would just keep it